src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -4,67 +4,196 @@
 ///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
 ///  to the pool.
+///
+///  There is significant effort to avoid retrieving the <see cref="ImmutableArray{T}.Builder"/>.
+///  For very small arrays of length 4 or less, the elements will be stored on the stack. If the array
+///  grows larger than 4 elements, a builder will be employed.
 /// </summary>
-internal ref partial struct PooledArrayBuilder<T>
+internal partial struct PooledArrayBuilder<T> : IDisposable
 {
-    private readonly ObjectPool<ImmutableArray<T>.Builder> _pool;
-    private readonly int? _capacity;
+    public static readonly PooledArrayBuilder<T> Empty = default;
+
+    /// <summary>
+    ///  The number of items that can be stored inline.
+    /// </summary>
+    private const int InlineCapacity = 4;
+
+    /// <summary>
+    ///  A builder used as storage when the number of items exceeds 4 elements.
+    /// </summary>
     private ImmutableArray<T>.Builder? _builder;
 
-    public PooledArrayBuilder()
-        : this(null, null)
+    /// <summary>
+    ///  An optional initial capacity for the builder.
+    /// </summary>
+    private int? _capacity;
+
+    private T _element0;
+    private T _element1;
+    private T _element2;
+    private T _element3;
+
+    /// <summary>
+    ///  The number of inline elements. Note that this value is only used when <see cref="_builder"/> is <see langword="null"/>.
+    /// </summary>
+    private int _inlineCount;
+
+    public PooledArrayBuilder(int? capacity = null)
     {
+        _capacity = capacity is > InlineCapacity ? capacity : null;
+        _element0 = default!;
+        _element1 = default!;
+        _element2 = default!;
+        _element3 = default!;
+        _inlineCount = 0;
     }
 
-    public PooledArrayBuilder(ObjectPool<ImmutableArray<T>.Builder>? pool = null, int? capacity = null)
+    private PooledArrayBuilder(in PooledArrayBuilder<T> builder)
     {
-        _pool = pool ?? ArrayBuilderPool<T>.Default;
-        _capacity = capacity;
+        // This is an intentional copy used to create an Enumerator.
+        this = builder;
     }
 
-    public readonly T this[int i]
+    public void Dispose()
     {
-        get
+        if (_builder is { } builder)
         {
-            if (_builder is null || Count <= i)
+            ArrayBuilderPool<T>.Default.Return(builder);
+            _builder = null;
+        }
+        else
+        {
+            for (var i = 0; i < _inlineCount; i++)
             {
-                throw new IndexOutOfRangeException();
+                ClearInlineElement(i);
             }
 
-            return _builder[i];
+            _inlineCount = 0;
         }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void ClearInlineElement(int index)
+    {
+        Debug.Assert(_inlineCount <= InlineCapacity);
+
+        // Clearing out an item makes it potentially available for garbage collection.
```

Comment:
If this struct is mostly on stack, this should not be needed according to comment in `TemporaryArray.Dispose`.

Reply:
This is needed here because this method is called in more places than just `Dispose`. For example, it is also called in `MoveInlineItemsToBuilder` where it is definitely necessary. However, you're right that it's unneeded in `Dispose` (for the same reason as `TemporaryArray<T>.Dispose`. I'll go ahead and update `Dispose` to stop clearing the elements and add a similar comment.